### PR TITLE
CSF: Deprecate displayName parameter

### PR DESCRIPTION
**Pull request by shilman**
Issue: #8756 #8507

## What I did

- [x] Map CSF `story.name` => story name
- [x] Remove `parameters.displayName` from CSF loading
- [x] Add `default.id` to CSF for story perma-linking (see #8808)
- [x] Update `MIGRATION.md`

## How to test

See updated snapshots

### CODE REVIEW

**Comment by Hypnosphi** [comment]
Thinks to check, based on what I encountered in #8508:
- [ ] Addon Storysource
- [ ] MDX stories

Please also update https://github.com/storybookjs/storybook/blob/master/docs/src/pages/formats/component-story-format/index.md#storybook-export-vs-name-handling

**Comment by tmeasday** [comment]
I don't see a deprecation for `displayName`? Otherwise seems reasonable.

**Comment by shilman** [comment]
I'd like to release this in 5.3 although it's technically a breaking change.

Here's the rationale:
- It's a much simpler way to do things
- The 5.2 CSF change was also a breaking change, and although this subjects existing 5.2 upgraders to two breaking changes (storyshots broke in 5.2 for people who upgraded to CSF, and will break again to fix the problem), I imagine that most people haven't upgraded yet, so the sooner we can get this out the better.

Apologies to semver purists. We're doing our best here!

**Review by mrmckeb — APPROVED** [review]

File: MIGRATION.md [hunk 1]
```diff
@@ -109,6 +110,12 @@ getStorybookUI({
 });
 ```
 
+### Deprecate displayName parameter
+
+In 5.2, the story parameter `displayName` was introduced as a publicly visible (but internal) API. Storybook's Component Story Format (CSF) loader used it to modify a story's display name independent of the story's `name`/`id` (which were coupled).
+
+In 5.3, the CSF loader decouples the story's `name`/`id`, which means that `displayName` is no longer necessary. Unfortunately, this is a breaking change for any code that uses the story `name` field. Storyshots relies on story `name`, and the appropriate mgiration is to simply update your snapshots. Apologies for the inconvenience!
```
Comment (tmeasday) [inline MIGRATION.md]:
```suggestion
In 5.3, the CSF loader decouples the story's `name`/`id`, which means that `displayName` is no longer necessary. Unfortunately, this is a breaking change for any code that uses the story `name` field. Storyshots relies on story `name`, and the appropriate migration is to simply update your snapshots. Apologies for the inconvenience!
```

File: lib/core/src/client/preview/start.js [hunk 2]
```diff
@@ -420,11 +420,12 @@ export default function start(render, { decorateStory } = {}) {
             `${kindName} => ${name || key}: story.parameters.decorators is deprecated; use story.decorators instead.`)();
           }
           const decoratorParams = decorators ? { decorators } : null;
-          const displayNameParams = name ? { displayName: name } : {};
-          kind.add(storyNameFromExport(key), storyFn, {
+          const exportName = storyNameFromExport(key);
+          const idParams = { __id: toId(kindName, exportName) };
```
Comment (tmeasday) [inline lib/core/src/client/preview/start.js]: Should we only pass this if `name` is defined?

File: lib/client-api/src/client_api.ts [hunk 3]
```diff
@@ -207,7 +207,7 @@ export default class ClientApi {
     api.add = (storyName, storyFn, parameters) => {
       hasAdded = true;
 
-      const id = toId(kind, storyName);
+      const id = (parameters && parameters.__id) || toId(kind, storyName);
```
Comment (tmeasday) [inline lib/client-api/src/client_api.ts]: NIT: just set `parameters = {}` above

**Review by tmeasday — CHANGES_REQUESTED** [review]
Coupla small things, LGTM otherwise

File: lib/core/src/client/preview/start.js (hunk 2 above)
Comment (shilman) [reply lib/core/src/client/preview/start.js]: Hmm could see arguments both ways. I think it's simpler to do the calculation here so it's all in one place.

File: lib/client-api/src/client_api.ts (hunk 3 above)
Comment (shilman) [reply lib/client-api/src/client_api.ts]: Fixed

**Pull request merged**